### PR TITLE
Add pre-commit config and a new workflow to run the pre-commit hooks on any PR

### LINKED ISSUE #502: Dev-Experience: Add a pre-commit GH workflow that will run the linting of files for any PR and potentially auto fix it
## Why do we need this ?

1. This is will improve the devexperience
2. Code standards will be maintained
3. No more manual lint checking or using just `black`(As black is integrated with pre-commt hooks)
4. If all agree, we can fix the lining in CI only(Debatable)

## Acceptance Criteria

- [x] Create a .pre-commit-config.yaml file and add the custom listing logic there
- [x] Create the GH workflow that will run the pre-commit action
- [ ] Format the entire codebase with pre-commit hooks

## Further Comments / References  

I worked on a project https://github.com/learningequality/kolibri, this uses a similar kind of workflow

### PR DESCRIPTION
**Pull request by thesujai**
## Linked Issue(s) 

Closes #502 

## Acceptance Criteria fulfillment

- [x] Create a .pre-commit-config.yaml file and add the custom listing logic there
- [x] Create the GH workflow that will run the pre-commit action
- [x] Format the entire codebase with pre-commit hooks

## Proposed changes (including videos or screenshots)

1. Change the scripts command in package.json of both node sub dir(cli/ and web-server/). As eslint should be ran with a proper path. Earlier running `yarn lint` resulted in nothing because paths were not configured
2. Introduced pre-commit to Run before committing anything. As of now, I have setup for only linitng and formatting files before committing. I can integrate more pre-commit hooks if you want one. But these are the ones I think are necessary.
3. Added a new GH workflow to run the pre-commit hooks.(Just saying it would have been a little simpler if both web-server/ and cli/ used the same version of node)(So I setup two node versions in the workflow)

## Further comments

1. I tested this with a test [PR](https://github.com/thesujai/middleware/pull/1). Though you see the hooks failed there, but I created that PR to see if the workflow setup runs correctly or not. It does work. 
2. Now After you review this PR, I will also push the formatted files(Not pushing now as reviewing would be gnarly for the maintainers)

### CODE REVIEW

**Comment by thesujai** [comment]
Just so you know we dont need a separate `black` workflow with this one.

Feel free to close it if the feature is not desired for now:)

**Comment by jayantbh** [comment]
Hey @thesujai, thanks for your contribution!

I'm curious about why `flake8` was introduced instead of just using `black`.

File: backend/dev-requirements.txt
Comment (thesujai) [inline backend/dev-requirements.txt:3]: Should probably bump up the pre-commit here as well

**Comment by jayantbh** [comment]
One of the checks failed. Interesting.
If you could look into sorting it out (our team will help), we'd love to accept this PR. 🤝

**Comment by thesujai** [comment]
Hi @jayantbh for a quick review
Flake8 does so much more than just lining the code(It checks for unused imports, also potential bugs!)
Like for eg the below code will be given a green flag by black:
```
import os
print("Hello")
``` 

But flake8 will throw an F401(Imported but not used) as the os module is imported but never used.
For now black ignores such best practices and focuses on PEP8 guidelines(Which don't include a F401 type thing!). So most error thrown by flake8 should be manually fixed(As no library I know of does things like this automatically)


> One of the checks failed

Yes, it is supposed to fail, I just introduced it with this PR. It will pass once I push all the formatted files(formatted by the pre-commit hooks). I am not pushing now as reviewing would be really irritating with so many lint-fixed files

**Comment by jayantbh** [comment]
Sounds good. I'd like @samad-yar-khan and @adnanhashmi09 to confirm if `flake8` plays well with our dev setup, so we can replace `black` with it.

**Comment by thesujai** [comment]
I am just bad with explaining my thoughts 

More like having flake8 along with black(as black is just the formatter, flake8 is just a checker). But both of these we will have as pre-commit hooks(with the pr introduces). 

Just the think to remove will be the black github action, as black will be ran as the last hook of pre-commit with the new pre-commit workflow.

**Comment by jayantbh** [comment]
@thesujai there's a conflict, and a failing lint test.

**Comment by thesujai** [comment]
@jayantbh It should work now. I rebased it with main.

Any changes in the upstream will again make the workflow to fail if this gets merged, so we have to make sure this is up to date with main before it is merged(it is now!)

**Comment by jayantbh** [comment]
Hey @thesujai, this PR is currently on pause because we're waiting for some other PRs to get merged to avoid rework on them. We'll take it up again in 1-2 days.
Apologies for the slow movement on this.

**Comment by thesujai** [comment]
Add the related docs as well(Pardon the grammar there 🙃 )

**Comment by jayantbh** [comment]
Thanks! I've triggered the workflows.

File: backend/analytics_server/mhq/api/teams.py
Comment (adnanhashmi09) [inline backend/analytics_server/mhq/api/teams.py:153]: Why `# noqa E501` here? These look like comments to suppress the linter.

**Review by adnanhashmi09 — COMMENTED** [review]
@thesujai 
1. why are SVGs also being linted? What is the reasoning behind it.
2. There are a lot of places in the python code where flake8 is being suppressed. Why?

File: backend/analytics_server/mhq/api/teams.py
Comment (thesujai) [reply backend/analytics_server/mhq/api/teams.py:153]: Just suppressing some linting rules. This one is for "line too long".
Seeing the codebase, I observe lots of places where lines are very long, refactoring all those makes no sense to me, as it would involve some code design change as well.

Not necessarily we  need need to follow all the standards of flake8.

**Comment by adnanhashmi09** [comment]
@thesujai Got it. Thanks for the clarifications. We should have a followup PR to have these comments removed. But for now it looks good to me.

**Review by adnanhashmi09 — APPROVED** [review]
LGTM

File: CONTRIBUTING.md
Comment (samad-yar-khan) [inline CONTRIBUTING.md:48]: @thesujai We need to make this doc better. Include a step to create virtual environment. before installing dependancies. If there is a step in doc, just link to it.

**Review by samad-yar-khan — COMMENTED** [review]
https://github.com/user-attachments/assets/969ca40a-5488-497d-8b36-8bd8e6120fd6

So, it tells me whats wrong but shouldn't the precommit hook fix these for me ?

**Review by samad-yar-khan — CHANGES_REQUESTED** [review]
Pls update the doc as mentioned above + check if the precommit can lint code on its own. 
Preferably, if someone is using /.dev.sh for dev, they are looking to change the code. See if we can get them to install the precommit at the top rather than potential misses and a lint commit at the end when they discover the precommit hook

**Comment by thesujai** [comment]
> So, it tells me whats wrong but shouldn't the precommit hook fix these for me ?

I can add [`autoflake`](https://github.com/PyCQA/autoflake?tab=readme-ov-file#using-pre-commit-hooks) for that

**Comment by jayantbh** [comment]
I missed this, but yeah a linter should absolutely try to fix as many issues automatically as it can.

**Comment by thesujai** [comment]
> Preferably, if someone is using /.dev.sh for dev, they are looking to change the code. See if we can get them to install the precommit at the top rather than potential misses and a lint commit at the end when they discover the precommit hook

I don't understand what you trying to say here. Let me know if the latest commit solves your concern!

**Comment by samad-yar-khan** [comment]
> > Preferably, if someone is using /.dev.sh for dev, they are looking to change the code. See if we can get them to install the precommit at the top rather than potential misses and a lint commit at the end when they discover the precommit hook
> 
> I don't understand what you trying to say here. Let me know if the latest commit solves your concern!

Hey @thesujai , that last commit solves a part of the problem of automatic linting with pre-commit hook. This still does not solve the effort of manually installing python dependancies just to run the precommit hook.
Currently our dev setup only requires the user to use the `./dev.sh` to develop using docker. All things needed by the dev setup should be a part of that script. This means, if I use dev.sh, it should automatically check if I need to install precommit hooks or any dependancies, install them at system level or in a virtual env and activate venv (in case of python) and let the dev build and commit without having to manually go and install these just to commit.

**Comment by thesujai** [comment]
@samad-yar-khan I updated the PR

**Comment by samad-yar-khan** [comment]
@thesujai I see the updated dev.sh, nice work on that. I don't have to install python packages now. But, its not working for me. Also, one of the linting workflows is failing.

https://github.com/user-attachments/assets/ebfcc4bd-330c-4484-963c-19c1ea1b2fb5

**Comment by thesujai** [comment]
@samad-yar-khan What node version were you using?

**Comment by samad-yar-khan** [comment]
Was previously using 18, now using 16.19.1

**Comment by thesujai** [comment]
From here it looks like your `cli/` had node modules installed, but not your `webserver/`.

We can configure dev.sh to also install the node_moduels(locally, not in a docker container), but this would mess up the current workflow where node_modules are getting installed into docker containers.

I suggest doing this but only adding eslint(and other linting tools) to the dev.sh workflow that will install node_modules(only with required tools) locally. But this would be extra work to keep maintaining the tools versioning in package.json as well as the dev.sh script.

Let me know if you have other ideas.

How I like to view things is that development is one thing and contributing is another.

**Comment by thesujai** [comment]
Just like in this workflow, I am installing the lining [tools](https://github.com/middlewarehq/middleware/pull/503/files#diff-ac5eead3f3ce4863c524fff031a87b7aecccb4a0493df087a4e1c704a1505036R38) for webserver/ then for cli/ [here](https://github.com/middlewarehq/middleware/pull/503/files#diff-ac5eead3f3ce4863c524fff031a87b7aecccb4a0493df087a4e1c704a1505036R47)

**Comment by thesujai** [comment]
> Can't we just use docker volumes to map the node_modules and python dependencies to the host's filesystem?

Possible, but there are some tradeoffs:

1. It might compromise the container’s isolation by exposing dependencies to the host. The major advantage of Docker is keeping dependencies isolated within the container.
2. Docker volumes can cause compatibility problems across different OSs, especially with node_modules. The docker base image isn't the same as host OS.

**Comment by adnanhashmi09** [comment]
> Possible, but there are some tradeoffs:
> 
> 1. It might compromise the container’s isolation by exposing dependencies to the host. The major advantage of Docker is keeping dependencies isolated within the container.
> 2. Docker volumes can cause compatibility problems across different OSs, especially with node_modules. The docker base image isn't the same as host OS.

Yeah makes sense. To run the linting process we would have to install node_moduels and python dependencies anyhow. Maybe we could remove it from the dev process and add to the contributing docs. This is making things a little messy in my opinion. 

@jayantbh Thoughts?

**Comment by adnanhashmi09** [comment]
@thesujai I'd say let's remove the installation of linters from the dev process and make this a part of `CONTRIBUTING.md`. We'd leave the installation of linters to the contributors and just give them steps how to do that.

**Comment by thesujai** [comment]
Clean now, rebased with main branch as well

File: CONTRIBUTING.md
Comment (adnanhashmi09) [inline CONTRIBUTING.md:27]:
```suggestion
1. Pre-commit should be installed as a dev dependency already. If not then Create a virual environment [venv](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/#create-and-use-virtual-environments) or [pyenv](https://github.com/pyenv/pyenv?tab=readme-ov-file#installation) and run the following command from project root dir:
```

File: CONTRIBUTING.md
Comment (adnanhashmi09) [inline CONTRIBUTING.md:43]:
```suggestion
    cd web-server
```

File: CONTRIBUTING.md
Comment (adnanhashmi09) [inline CONTRIBUTING.md:40]: No need to install `eslint-config-next` for cli as it doesn't depend on nextjs. And running this command makes changes in the `package.json` for `cli/`. It'd be great if you could push those changes in this PR itself.

**Review by adnanhashmi09 — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (thesujai) [reply CONTRIBUTING.md]: Ah good catch!

**Comment by thesujai** [comment]
@adnanhashmi09 should work?

**Comment by jayantbh** [comment]
Amazing. This was a trip. 🚀🚀🚀
Great work @thesujai 👏

**Pull request merged**
